File: extract/sc_extract/poses.py
"""Retarget poses out of the game's own animation data.

Star Citizen keeps character animation in CryEngine ``.dba`` databases named by
the skeleton's ``.chrparams``. For a bare-handed human they sit under
``Animations/Characters/Human/<skeleton>/weapons/no_weapon/locomotion/``.

StarBreaker parses both the ``.chr`` skeleton and the animation formats but
exposes neither on its CLI, so ``tools/anim-dump`` is a small shim over its
``starbreaker-3d`` crate.

**Why this is a retarget and not a copy.** A clip stores absolute *local*
rotations in the animation rig's own bone frames. Our skeleton reached glTF via
cgf-converter, Collada, Blender and the glTF exporter, and its local frames no
longer match, so applying those rotations directly lays the character on its
back.

Absolute *world* orientation does not transfer either: it only works where the
two rigs' bone axes happen to agree, which they do for the spine and legs and
do not for the arms, whose bind differs. Arms ended up pointing at the ceiling.

What does transfer is the **delta from each rig's own bind pose**:
``delta = world_clip * inverse(world_bind)``. That says "rotate this bone by
however far the animation moves it from rest", which is independent of either
rig's axis conventions. The viewer applies it to our own rest orientation.
Bone lengths stay ours, so the pose adapts to our proportions.

The clip's world space has up along -Y and forward along +Z, which reaches glTF
through a 180 degree rotation about X. That was read off the data: a standing
clip puts the head 1.70 from the floor along -Y, and a crouch puts the knee
forward at +Z.
"""

from __future__ import annotations

import json
import logging
import subprocess
from dataclasses import dataclass
from pathlib import Path

from .catalog import SKELETON_ROOTS
from .config import REPO_ROOT, Settings
from .tools import ToolError

log = logging.getLogger(__name__)

ANIM_DUMP = REPO_ROOT / "tools" / "anim-dump" / "target" / "release" / "anim-dump"
LOCOMOTION = "Animations/Characters/Human/{skeleton}/weapons/no_weapon/locomotion"
SKELETON_CHR = {
    "male": "Objects/Characters/Human/male_v7/export/bhm_skeleton_v7.chr",
    "female": "Objects/Characters/Human/female_v2/export/bhf_skeleton_v2.chr",
}

Quat = tuple[float, float, float, float]  # w, x, y, z
Vec3 = tuple[float, float, float]


@dataclass(frozen=True)
class PoseSpec:
    name: str
    label: str
    database: str
    clip: str


# Chosen for their final frame. Clips suffixed ``_add`` are additive deltas
# layered on a base at runtime and are no use as a standalone pose.
POSES: tuple[PoseSpec, ...] = (
    PoseSpec("idle", "Idle", "stand.dba", "nw_stand_idle_turn360_planted"),
    PoseSpec("crouch", "Crouch", "crouch.dba", "nw_neutral_crouch_idle.caf"),
)


def qmul(a: Quat, b: Quat) -> Quat:
    w1, x1, y1, z1 = a
    w2, x2, y2, z2 = b
    return (
        w1 * w2 - x1 * x2 - y1 * y2 - z1 * z2,
        w1 * x2 + x1 * w2 + y1 * z2 - z1 * y2,
        w1 * y2 - x1 * z2 + y1 * w2 + z1 * x2,
        w1 * z2 + x1 * y2 - y1 * x2 + z1 * w2,
    )


def qrot(q: Quat, v: Vec3) -> Vec3:
    w, x, y, z = q
    vx, vy, vz = v
    tx = 2 * (y * vz - z * vy)
    ty = 2 * (z * vx - x * vz)
    tz = 2 * (x * vy - y * vx)
    return (
        vx + w * tx + (y * tz - z * ty),
        vy + w * ty + (z * tx - x * tz),
        vz + w * tz + (x * ty - y * tx),
    )


def to_gltf_quat(q: Quat) -> list[float]:
    """Clip world space to glTF, as xyzw. A 180 degree turn about X."""
    w, x, y, z = q
    return [x, -y, -z, w]


def to_gltf_position(v: Vec3) -> list[float]:
    x, y, z = v
    return [x, -y, -z]


def run_dump(args: list[str]) -> str:
    if not ANIM_DUMP.is_file():
        raise ToolError(
            "anim-dump is not built. Run:\n"
            "  cargo build --release --manifest-path tools/anim-dump/Cargo.toml\n"
            "It needs the StarBreaker clone that tools/build.sh creates."
        )
    proc = subprocess.run(  # noqa: S603
        [str(ANIM_DUMP), *args], capture_output=True, text=True, check=False
    )
    if proc.returncode != 0:
        raise ToolError(f"anim-dump {args[0]} failed: {proc.stderr.strip()[:300]}")
    return proc.stdout


def qconj(q: Quat) -> Quat:
    w, x, y, z = q
    return (w, -x, -y, -z)


def forward_kinematics(bind: list[dict], clip: dict) -> dict[str, dict]:
    """Per-bone rotation delta from the bind pose, in glTF axes.

    Bones the clip does not animate keep their bind local transform, so their
    delta comes out as identity.
    """
    world_rotation: list[Quat] = [(1.0, 0.0, 0.0, 0.0)] * len(bind)
    world_position: list[Vec3] = [(0.0, 0.0, 0.0)] * len(bind)

    for i, bone in enumerate(bind):
        local_rotation: Quat = tuple(bone["local_rotation"])  # type: ignore[assignment]
        local_position: Vec3 = tuple(bone["local_position"])  # type: ignore[assignment]

        entry = clip.get(bone["name"])
        if entry:
            local_rotation = tuple(entry["rotation"])  # type: ignore[assignment]
            if "position" in entry:
                local_position = tuple(entry["position"])  # type: ignore[assignment]

        parent = bone["parent"]
        if parent is None:
            world_rotation[i] = local_rotation
            world_position[i] = local_position
        else:
            world_rotation[i] = qmul(world_rotation[parent], local_rotation)
            offset = qrot(world_rotation[parent], local_position)
            world_position[i] = tuple(
                world_position[parent][k] + offset[k] for k in range(3)
            )

    out: dict[str, dict] = {}
    for i, bone in enumerate(bind):
        bind_world: Quat = tuple(bone["world_rotation"])  # type: ignore[assignment]
        delta = qmul(world_rotation[i], qconj(bind_world))
        out[bone["name"]] = {
            "delta": to_gltf_quat(delta),
            "position": to_gltf_position(world_position[i]),
        }
    return out


def build(settings: Settings, *, skeleton: str | None = None) -> Path:
    """Write ``data/out/poses.json`` from the game's animation databases."""
    skeleton = skeleton or settings.skeleton
    skeleton_json = settings.base_dir() / f"{skeleton}.skeleton.json"
    if not skeleton_json.is_file():
        raise ToolError(f"no skeleton at {skeleton_json}; run `scx rig` first")
    known = set(json.loads(skeleton_json.read_text()).get("bones", []))

    chr_path = settings.raw_dir / "Data" / SKELETON_CHR[skeleton]
    if not chr_path.is_file():
        raise ToolError(f"no skeleton at {chr_path}; extract it first")
    bind = json.loads(run_dump(["bind", str(chr_path)]))

    directory = SKELETON_ROOTS.get(skeleton, skeleton)
    root = settings.raw_dir / "Data" / LOCOMOTION.format(skeleton=directory)

    out: dict[str, dict] = {}
    for spec in POSES:
        database = root / spec.database
        if not database.is_file():
            raise ToolError(f"no animation database at {database}")
        clip = json.loads(
            run_dump(["pose", str(database), spec.clip, str(skeleton_json)])
        )
        world = forward_kinematics(bind, clip)
        bones = {name: entry for name, entry in world.items() if name in known}
        out[spec.name] = {"label": spec.label, "clip": spec.clip, "bones": bones}
        log.info("%s: %d bones from %s", spec.name, len(bones), spec.clip)

    target = settings.out_dir / "poses.json"
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(json.dumps(out, indent=1) + "\n")
    return target

File: web/core/src/poses.rs
//! Retargeting a game animation clip onto the canonical armature.
//!
//! Three approaches were tried in the Python and two of them fail, so the one
//! here is not a choice of convenience:
//!
//! 1. *Copy the clip's local rotations.* Fails. A clip stores absolute local
//!    rotations in the animation rig's bone frames, and ours no longer match
//!    after Collada, Blender and the glTF exporter. The character ends up on
//!    its back, and no axis permutation fixes it -- seven were tried.
//! 2. *Copy world orientations,* obtained by running forward kinematics over
//!    the `.chr` hierarchy. Spine and legs land correctly and the **arms point
//!    at the ceiling**: absolute orientation only transfers where the two rigs'
//!    bone axes agree, and for arms they do not.
//! 3. *Transfer the delta from each rig's own bind pose*,
//!    `delta = world_clip · inverse(world_bind)`. This works, because "rotate
//!    this bone by however far the animation moves it" needs no agreement about
//!    axes at all. Bone lengths stay ours, so the pose adapts to our
//!    proportions.
//!
//! Two details the data dictates. The clip's world space has **up along -Y and
//! forward along +Z**, reaching glTF through a 180-degree rotation about X.
//! And the armature **root is skipped** by the caller: it carries the clip's own
//! world placement, which otherwise drags the whole body a metre sideways.

/// A quaternion as `[w, x, y, z]`, matching the archive's own ordering.
pub type Quat = [f32; 4];
pub type Vec3 = [f32; 3];

/// One bone of the bind pose, as the `.chr` stores it.
#[derive(Debug, Clone)]
pub struct BindBone {
    pub name: String,
    pub parent: Option<usize>,
    pub local_rotation: Quat,
    pub local_position: Vec3,
    pub world_rotation: Quat,
}

/// What a clip says about one bone at the frame being sampled.
#[derive(Debug, Clone, Default)]
pub struct Sample {
    pub rotation: Option<Quat>,
    pub position: Option<Vec3>,
}

/// The retargeted pose for one bone, in glTF axes.
#[derive(Debug, Clone, PartialEq)]
pub struct Posed {
    /// Rotation delta as glTF `[x, y, z, w]`.
    pub delta: [f32; 4],
    pub position: Vec3,
}

pub fn qmul(a: Quat, b: Quat) -> Quat {
    let (aw, ax, ay, az) = (a[0], a[1], a[2], a[3]);
    let (bw, bx, by, bz) = (b[0], b[1], b[2], b[3]);
    [
        aw * bw - ax * bx - ay * by - az * bz,
        aw * bx + ax * bw + ay * bz - az * by,
        aw * by - ax * bz + ay * bw + az * bx,
        aw * bz + ax * by - ay * bx + az * bw,
    ]
}

pub fn qconj(q: Quat) -> Quat {
    [q[0], -q[1], -q[2], -q[3]]
}

/// Rotate a vector by a quaternion.
pub fn qrot(q: Quat, v: Vec3) -> Vec3 {
    let t = [
        2.0 * (q[2] * v[2] - q[3] * v[1]),
        2.0 * (q[3] * v[0] - q[1] * v[2]),
        2.0 * (q[1] * v[1] - q[2] * v[0]),
    ];
    [
        v[0] + q[0] * t[0] + q[2] * t[2] - q[3] * t[1],
        v[1] + q[0] * t[1] + q[3] * t[0] - q[1] * t[2],
        v[2] + q[0] * t[2] + q[1] * t[1] - q[2] * t[0],
    ]
}

/// Clip world space to glTF, as xyzw. A 180-degree turn about X.
pub fn to_gltf_quat(q: Quat) -> [f32; 4] {
    [q[1], -q[2], -q[3], q[0]]
}

pub fn to_gltf_position(v: Vec3) -> Vec3 {
    [v[0], -v[1], -v[2]]
}

/// Run forward kinematics over the bind hierarchy, substituting whatever the
/// clip animates, and return each bone's delta from its own bind pose.
///
/// Bones the clip does not animate keep their bind local transform, so their
/// delta comes out as identity -- which is what makes a clip that touches 145
/// of 220 bones leave the other 75 alone rather than resetting them.
pub fn forward_kinematics(
    bind: &[BindBone],
    sample_of: impl Fn(&str) -> Option<Sample>,
) -> Vec<(String, Posed)> {
    let mut world_rotation: Vec<Quat> = vec![[1.0, 0.0, 0.0, 0.0]; bind.len()];
    let mut world_position: Vec<Vec3> = vec![[0.0, 0.0, 0.0]; bind.len()];

    for (i, bone) in bind.iter().enumerate() {
        let mut local_rotation = bone.local_rotation;
        let mut local_position = bone.local_position;
        if let Some(sample) = sample_of(&bone.name) {
            if let Some(r) = sample.rotation {
                local_rotation = r;
            }
            if let Some(p) = sample.position {
                local_position = p;
            }
        }

        match bone.parent {
            None => {
                world_rotation[i] = local_rotation;
                world_position[i] = local_position;
            }
            Some(parent) => {
                world_rotation[i] = qmul(world_rotation[parent], local_rotation);
                let offset = qrot(world_rotation[parent], local_position);
                world_position[i] = [
                    world_position[parent][0] + offset[0],
                    world_position[parent][1] + offset[1],
                    world_position[parent][2] + offset[2],
                ];
            }
        }
    }

    bind.iter()
        .enumerate()
        .map(|(i, bone)| {
            let delta = qmul(world_rotation[i], qconj(bone.world_rotation));
            (
                bone.name.clone(),
                Posed {
                    delta: to_gltf_quat(delta),
                    position: to_gltf_position(world_position[i]),
                },
            )
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    const ID: Quat = [1.0, 0.0, 0.0, 0.0];

    fn bone(name: &str, parent: Option<usize>, local: Quat, world: Quat) -> BindBone {
        BindBone {
            name: name.into(),
            parent,
            local_rotation: local,
            local_position: [0.0, 0.0, 0.0],
            world_rotation: world,
        }
    }

    #[test]
    fn an_unanimated_bone_gets_an_identity_delta() {
        // The reason a clip touching 145 of 220 bones leaves the rest alone.
        let bind = vec![bone("World", None, ID, ID), bone("Hips", Some(0), ID, ID)];
        let out = forward_kinematics(&bind, |_| None);
        for (name, posed) in &out {
            assert_eq!(posed.delta, [0.0, -0.0, -0.0, 1.0], "{name} should not move");
        }
    }

    #[test]
    fn the_delta_is_measured_from_the_bone_s_own_bind_pose() {
        // A quarter turn about Z, as the clip would store it.
        let turn: Quat = [0.70710677, 0.0, 0.0, 0.70710677];
        // The bind already has that turn, so animating to it must be a no-op.
        let bind = vec![bone("World", None, turn, turn)];
        let out = forward_kinematics(&bind, |_| {
            Some(Sample { rotation: Some(turn), position: None })
        });
        let d = out[0].1.delta;
        assert!((d[3] - 1.0).abs() < 1e-5, "identity delta, got {d:?}");
    }

    #[test]
    fn rotation_accumulates_down_the_chain() {
        let turn: Quat = [0.70710677, 0.0, 0.0, 0.70710677];
        let bind = vec![bone("World", None, ID, ID), bone("Hips", Some(0), ID, ID)];
        // Turning the parent must move the child too.
        let out = forward_kinematics(&bind, |n| {
            (n == "World").then_some(Sample { rotation: Some(turn), position: None })
        });
        let child = out[1].1.delta;
        assert!((child[3] - 0.70710677).abs() < 1e-5, "child inherits the turn: {child:?}");
    }

    #[test]
    fn gltf_conversion_is_a_180_about_x() {
        // [w,x,y,z] in, [x,y,z,w] out, with y and z negated.
        assert_eq!(to_gltf_quat([1.0, 2.0, 3.0, 4.0]), [2.0, -3.0, -4.0, 1.0]);
        assert_eq!(to_gltf_position([1.0, 2.0, 3.0]), [1.0, -2.0, -3.0]);
    }
}
